### ISSUE TITLE
[question] pre-built binary that includes an executable

### ISSUE THREAD

**Issue by dwswingle** [conan-io/conan#17202]
### What is your question?

I have a pre-built binary recipe that has a bin directory that needs to get added to the PATH of a consuming package. I noticed this when looking at CMake and/or Conan recipes:

```
 # Needed for compatibility with v1.x - Remove when 2.0 becomes the default
        bindir = os.path.join(self.package_folder, "bin")
        self.output.info(f"Appending PATH environment variable: {bindir}")
        self.env_info.PATH.append(bindir)
```

What is the "new" way to do this with Conan 2? thanks!

### Have you read the CONTRIBUTING guide?

- [x] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @dwswingle 
Thanks for your question.

Conan 2 already includes the ``cpp_info.bindirs`` of the dependencies in the PATH of the consumers by default. So all that is necessary is:
- Make sure that the executables are in the "bin" folder of the package (or re-define the default ``self.cpp_info.bindirs``)
- Use the ``VirtualBuildEnv`` and ``VirtualRunEnv`` generators (these are also activated implicitly by default)
- When the consumer calls ``conan install`` it will generate ``conanbuild`` and ``conanrun`` scripts. Activating them for the build (to activate PATH of tool-requires) or for executing (to activate the PATH of requires) respectively, will define the PATH variable

Please try that and let us know.


As a note, if a package wants to put some env-var explicitly into the consumers, the interface is in ``package_info()`` via the ``self.buildenv_info`` and ``self.runenv_info`` for defining the env-var for build or run time of the consumer respecitvely. But not needed for the PATH one (or LD_LIBRARY_PATH, DYLD_LIBRARY_PATH, those are also automatic)

**Comment by dwswingle** [comment]
if that is the case, wouldn't this be good enough?

**conanfile.py**
```
    def package(self):
        copy(self, "*.exe", self.build_folder, os.path.join(self.package_folder, "bin"))
```

but if i try this in the test_package's conanfile.py, it fails (test.exe can't be found)

```
    def test(self):
        if can_run(self):
            self.run("test.exe --version")
```

**Comment by memsharded** [comment]
If the package is a ``tool_requires`` it should work automatically.
If not, it is necessary to do ``self.run("test.exe --version", env="conanrun")`` 

(In general the default for ``self.run()`` is "building" things in the build-time, this is why it defaults to ``env="conanbuild"``.

**Comment by dwswingle** [comment]
I feel like the problem I'm facing is that I'm trying to write a test_package for packaged pre-built binaries. All the examples of test_packages are generally written with a compiled CMake project in mind (something like google test, protobuffer, etc.). does that make sense? This is what I'm trying to follow: 

https://docs.conan.io/2/tutorial/creating_packages/other_types_of_packages/package_prebuilt_binaries.html#downloading-and-packaging-pre-built-binaries

My use case has headers, libraries, a license, and executables all prebuilt. the executable needs to consume a license when it runs.

**Comment by dwswingle** [comment]
Order of operation from my perspective

- tarball is sitting in artifactory
- the conan recipe extracts the tarball's headers/libraries/license/executable
- the test_package is a small cmake project that does a find_package on the conan package
- the test_package needs to be able to invoke the precompiled executable and pass in the license file to run successfully

**Comment by memsharded** [comment]
> I feel like the problem I'm facing is that I'm trying to write a test_package for packaged pre-built binaries. All the examples of test_packages are generally written with a compiled CMake project in mind (something like google test, protobuffer, etc.). does that make sense? This is what I'm trying to follow:

``test_package`` is completely independent of the build system (or the lack thereof) of the package. It is even very common to have packages built with other build systems, but still the ``test_package`` is done with CMake.
Same applies for packaging pre-compiled libraries, the ``test_package`` would still be independent. In fact, the ``conan export-pkg`` in Conan 2, that packages pre-compiled binaries, also automatically execute the ``test_package``.

> the test_package needs to be able to invoke the precompiled executable and pass in the license file to run successfully

Why? The ``test_package`` should do the minimal check to see if the package was correctly created, it is not intended to check any functionality.

If you are using executables from the package, the way to do it is by having 2 requirements, a regular ``requires`` (for the libraries) and a ``tool_requires`` (for the exes). Read https://docs.conan.io/2/examples/graph/tool_requires/using_protobuf.html

the issue is that doing that in ``test_package`` is not that straightforward, but it is doable.

**Comment by dwswingle** [comment]
appreciate the response. the executable that is prebuilt generates code so the protobuf example is great. will let you know if i need more assistance.

**Comment by dwswingle** [comment]
so I think I made some progress from looking at protobuffer's test package. I'm currently stuck on 

`CMake Error: CMake was unable to find a build program corresponding to "Ninja".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.`

when my test package is trying to configure. here is my current test_package's conanfile.py:

```
import os

from conan import ConanFile
from conan.tools.cmake import CMake, cmake_layout, CMakeDeps, CMakeToolchain
from conan.tools.build import can_run


class CoreDXTestConan(ConanFile):
    settings = "os", "compiler", "build_type", "arch"
    package_type = "shared-library"
    #generators = "CMakeDeps", "CMakeToolchain"

    def requirements(self):
        self.requires(self.tested_reference_str)

    def generate(self):
        deps = CMakeDeps(self)
        deps.check_components_exist = True
        deps.generate()
        tc = CMakeToolchain(self, generator="Ninja")
        tc.user_presets_path = 'ConanPresets.json'
        tc.generate()        

    def build(self):
        cmake = CMake(self)
        cmake.configure()
        cmake.build()

    def layout(self):
        cmake_layout(self, generator="Ninja")

    def test(self):
        if can_run(self):
            cmd = os.path.join(self.cpp.build.bindir, "example")
            self.run(cmd, env="conanrun")
```

**Comment by dwswingle** [comment]
would 
`self.cpp_info.resdirs = []`
be a reasonable attribute to point for a license file?

**Comment by memsharded** [comment]
> self.cpp_info.resdirs = []

Not really. The ``resdirs`` are more intended to be "compilation resources", more like assets that are involved in the build process.

In practice, the license files are located by convention, mostly with standard filenames or inside standard folders.

**Comment by dwswingle** [comment]
Understood. I may just be missing something or confused at this point.I am trying to call a precompiled binary (that requires a license file to be passed in when calling it) under my test_package's CMakeLists.txt to ensure consumers will be able to to do the same. I'm looking at all the generated CMake stuff in the test_packages' build folder. I'm trying to figure out what variable(s) I could use to pass in to an "add_custom_command".

As an example, this is how the binary and license were being called before
```
    add_custom_command (
        OUTPUT ${file_list}
        COMMAND ${CoreDX_TOP}/coredx_ddl -i p -l cpp11 -f ${idl_path}/${idl_name}.idl -d ${CMAKE_CURRENT_BINARY_DIR}/dds_product_files/${idl_name} -L ${COREDX_LICENSE_FILE} -I ${CMAKE_CURRENT_BINARY_DIR}/dds_product_files
        DEPENDS ${idl_path}/${idl_name}.idl
        )
```

where "CoreDX_TOP" and "COREDX_LICENSE_FILE" were hardcoded locations on a server previously.

I am successfully packaging the includes/libraries/binary/licenses in the package, just not sure how to access the binary/licenses in particular. includes/libraries are straight forward and already have that working in the test_package. thanks!

**Comment by memsharded** [comment]
Find the attached code: [src.zip](https://github.com/user-attachments/files/17639866/src.zip)

Some important points:

- The ``test_package`` contains a ``self.tool_requires(self.tested_reference_str)``, as it is using a tool from the dependency, that must be in the "build" context
- I made sure to add in the package a ``license.txt`` file, and ``package()`` it with ``copy(self, "license.txt", src=self.source_folder, dst=self.package_folder)``
- The ``main.cpp`` receives a command line argument with a path to the license, and reads and prints it, to guarantee that it is fine. It also generates some .c source code, just for the test ``add_custom_command``

In the ``test_package``:
- ``tc.cache_variables["MYPKG_PKG_FOLDER"] = self.dependencies.build["mypkg"].package_folder`` to pass the package folder. Note the ``dependencies.build``

Please let me know if this helps

**Comment by memsharded** [comment]
Hi @dwswingle 

Any feedback of the above code? Did it respond the questions? Thanks for the feedback!

**Comment by dwswingle** [comment]
The first time I looked at it I still had questions. Let me double check this week and get back to you.

**Comment by memsharded** [comment]
I'd also like to comment, that the new incubating CMakeDeps generator allows definition of ``cpp_info.exe`` that create CMake targets on the consumer side.

**Comment by dwswingle** [comment]
what version of Conan was this incorporated into?

**Comment by memsharded** [comment]
It is not a full fledged feature, it is "incubating", behind a feature flag, see https://docs.conan.io/2/incubating.html. But it is already there. Only the latest Conan release is recommended for testing this feature.

**Comment by dwswingle** [comment]
I am good to close this issue. I will look into that new feature. Appreciate all your help on this and the example code

**Comment by memsharded** [comment]
Thanks to you for the feedback!
